[RoeReRe] iP

RoeReRe opened the pull request:
# RoeBot
> “Now I am become Death, the destroyer of worlds.” – J. Robert Oppenheimer ([source](https://www.wired.co.uk/article/manhattan-project-robert-oppenheimer))

RoeBot frees your mind of having to remember things you need to do. It's,
- text-based
- useless
- ~~FAST~~ to use

All you need to do this,
1. download it from [here](https://github.com/RoeReRe/ip/releases)
2. add yours tasks
3. let it manage tasks for you :shipit:
4. uninstall because its useless!

And it is **FREE**! _(But I accept PayNow donations)_

Features:
- [x] Managing tasks
- [ ] Reminders (coming in 2030)
- [ ] Finding a cure for cancer (coming in 2050)
---
If you Java programmer, you can use it to practice Java too. Here's the `main` method:

```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(Launcher.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

RoeReRe committed:
Initial skeleton

RoeReRe committed:
Added echo function

RoeReRe committed:
Added feature to add and list tasks

RoeReRe committed:
Added ability to mark and unmark tasks

RoeReRe committed:
Added todos, deadlines and events

RoeReRe committed:
test with io redirection technique

RoeReRe committed:
Added error handling and retested with io technique

RoeReRe committed:
Added feature to delete tasks

RoeReRe committed:
Added javadoc to public methods

RoeReRe committed:
Edited implementation of prompt to prevent overflow

RoeReRe committed:
Add local database for tasks

RoeReRe committed:
Merge Level 7

RoeReRe committed:
Change task date representation to java LocalDataTime

RoeReRe committed:
Merge Level 8

RoeReRe committed:
Refactored the main Duke logic into additional classes

RoeReRe committed:
Divide classes into package

RoeReRe committed:
Merge remote-tracking branch 'origin/add-gradle-support'

The project currently is lacking Gradle support.

Adding Gradle would expedite JUnit tests.

RoeReRe committed:
Set up gradle for project

RoeReRe committed:
Add Junit tests for Todo and Deadline

RoeReRe committed:
Add more JavaDoc comments

RoeReRe committed:
Edited code to comply with coding standard

RoeReRe committed:
Add Find feature for tasks

RoeReRe committed:
Merge origin/branch-A-JavaDoc

RoeReRe committed:
Merge origin/branch-A-CodingStandard

RoeReRe committed:
Merge origin/branch-Level-9

RoeReRe committed:
Change UI from text-based to GUI

RoeReRe committed:
Revert "Change UI from text-based to GUI"

This reverts commit 7190626e4ff93b784e588720567fc9acbf5da476.
The reason is due to fast-forward which is not in line with the iP
specs.

RoeReRe committed:
Edit CSS file

RoeReRe committed:
Merge branch Level-10

RoeReRe committed:
Revert "Merge branch Level-10"

This reverts commit 60c0a58cda03057a78bad01e58f8c2aa5563bb22, reversing
changes made to 8c536e12ad53f3a90469defac0bb8d7e3af8fc20.

RoeReRe committed:
Merge branch 'master' into branch-Level-10

RoeReRe committed:
Merge branch 'branch-Level-10'

waseemingly commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    public abstract String toDatabaseRepresentation();
+
+    public void markAsDone() {
```
I like how you used a verb to name your methods.

waseemingly commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Event represents a task with a start and end date.
+ */
+public class Event extends Task {
```
I like how you used nouns to name your classes.

waseemingly commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public abstract class Task {
+    private String taskName;
+    private boolean isDone;
```
I like how your boolean variable is accompanied by an is prefix.

waseemingly commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,40 @@
+public class Duke {
+
+    public Storage storage;
+    protected TaskList tasks;
```
I like how you used plural for tasks as it is a multi-valued variable.

waseemingly commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,169 @@
+    /**
+     * Creates a local text file for saving the lists of tasks if it does not exist.
+     */
+    public void createDatabase() {
```
I like how your method is named such that it explains its function.

waseemingly reviewed (commented):
Overall, the naming conventions is well-adhered to and there is proper naming for classes. methods and variables.

jrchoo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,119 @@
+                dukeBot.tasks.findTasks(inputInfo);
+                break;
+            default:
+                throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
```
Nice use of enums and switch statement to handle the various commands.

jrchoo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,169 @@
+     *
+     * @param taskList The list containing the tasks to be saved.
+     */
+    public void writeToDatabase(ArrayList<Task> taskList) {
```
Methods for operations on the database are written as verbs, clear and easy to understand without having to read the JavaDoc comment.

jrchoo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,169 @@
+        }
+    }
+
+    private static Task stringToTask(String line) throws DukeException {
+        try {
+            Task newTask;
+            String[] taskInfo = line.split(" \\| ");
+
+            switch (taskInfo[0]) {
+                case "T":
+                    newTask = new Todo(taskInfo[2]);
+                    break;
+                case "D":
+                    newTask = new Deadline(taskInfo[2], taskInfo[3]);
+                    break;
+                case "E":
+                    newTask = new Event(taskInfo[2], taskInfo[3], taskInfo[4]);
+                    break;
+                default:
+                    throw new DukeException("Database is corrupted.");
+            }
+
+            if (taskInfo[1].equals("1")) {
+                newTask.markAsDone();
+            }
+
+            return newTask;
+        } catch (DukeException e) {
+            throw e;
+        }
+    }
```
Interesting to see that the creation of Task and its subclasses were handled in Storage. Maybe this was part of an old implementation so it may be a good idea to remove it to prevent confusion?

jrchoo commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    @Override
+    public String toDatabaseRepresentation() {
+        return String.format("D | %s | %s | %s",
+            isDone() ? 1 : 0, getTaskName(), Storage.dateToDatabaseRepresentation(this.deadline));
+    }
+
+    @Override
+    public String toString() {
+        return String.format("[D]%s (by: %s)", super.toString(), Storage.dateToString(this.deadline));
+    }
+}
```
I like that your string representation methods for Todo, Deadline and Event use String.format(). This looks a lot more elegant than hardcoding and something that I can try to do in my iP.

jrchoo reviewed (commented):
LGTM.

RoeReRe committed:
Add Java assertions

There are several methods that could use some extra layers of check to
prevent errors from causing damage.

Adding Java assertions would prevent these key errors from proceeding.

Let's add assertions to Storage and TaskList to prevent any corruption
to the database.

RoeReRe committed:
Improve code quality

The current codebase contains old code that is no longer in used.

Removing these unused classes would keep the repository clean and easy
to follow.

Let's improve the readability of the codebase.

RoeReRe committed:
Merge pull request #1 from RoeReRe/branch-A-Assertions

Add Java assertions

RoeReRe committed:
Merge pull request #2 from RoeReRe/branch-A-CodeQuality

Improve code quality

RoeReRe committed:
Add sort feature for tasks

The list function is very unreadable as the tasks are not in any sorted
order

By adding a sort function, we can sort Deadlines and Events by
chronological order

Let's keep the ToDos at the front and order the rest by date to make the
list more readable

RoeReRe committed:
Merge branch 'branch-C-Sort'

The master branch is lacking a sort function for tasks that is present
in branch-C-Sort

Let's merge the branch to improve the functionality of the bot

RoeReRe committed:
Add user guide and screenshot

RoeReRe committed:
Update README.md
